### PR TITLE
feat: add mods to player.last_np

### LINKED ISSUE #478: Add mods to beatmap.last_np
The `last_np` field is used for keeping context of what the player's last /np'ed.

We currently parse the mods out of the /np in regex:
https://github.com/osuAkatsuki/bancho.py/blob/a043f12e4a77f83faa180938214e02f79333872d/app/api/domains/cho.py#L74-L79

But we do not store this in the beatmap.last_np field after parsing from regex:
1) https://github.com/osuAkatsuki/bancho.py/blob/a043f12e4a77f83faa180938214e02f79333872d/app/api/domains/cho.py#L1226-L1230
2) https://github.com/osuAkatsuki/bancho.py/blob/a043f12e4a77f83faa180938214e02f79333872d/app/api/domains/cho.py#L402-L406

This data is useful so it can then later be used for e.g. pp calculation based on context.

### PR DESCRIPTION
**Pull request by minisbett**
# Describe your changes

Adds the mods, which are already being parsed from the /np regex, to the player.last_np object.

## Related Issues / Projects

Resolves #478 

## Checklist
- [X] I've manually tested my code
- [X] The changes pass pre-commit checks (`make lint`)
- [X] The changes follow coding style

### CODE REVIEW

**Comment by cmyui** [comment]
player.last_np is typed here as well - needs update

https://github.com/osuAkatsuki/bancho.py/blob/3a629386f1e8572b550becbd1f2558b1bbf23f4a/app/objects/player.py#L143-L146

**Comment by cmyui** [comment]
hmm mypy is not running on this pr - maybe cause it's on another repo? not sure

**Comment by minisbett** [comment]
> player.last_np is typed here as well - needs update
> 
> https://github.com/osuAkatsuki/bancho.py/blob/3a629386f1e8572b550becbd1f2558b1bbf23f4a/app/objects/player.py#L143-L146

oh shit

**Comment by tsunyoku** [comment]
> hmm mypy is not running on this pr - maybe cause it's on another repo? not sure
> 
> 
> 
> cc @tsunyoku 

just checked, you limited the workflow to the `push` action which is indeed restricted to pushes on any branch in the main repo.

you can copy the rules from the other workflows - adding a `pull_request` action should fix it

**Comment by cmyui** [comment]
> adding a pull_request action should fix it

nice let me try ty

File: app/api/domains/cho.py [hunk 1]
```diff
@@ -1217,9 +1224,16 @@ async def handle(self, player: Player) -> None:
...
+                        # parse the mods from regex
+                        if r_match["mods"] is not None:
+                            mods = Mods.from_np(r_match["mods"][1:], mode_vn)
+                        else:
+                            mods = None
+
                         player.last_np = {
                             "bmap": bmap,
                             "mode_vn": mode_vn,
+                            "mods": mods,
```
Comment (cmyui) [inline app/api/domains/cho.py:1236]: mods can be assigned as None here so the type of `LastNp.mods` should be `Mods | None`

if u use vs code, you can set the `type checking mode` setting to `"basic"` to get type checking in your editor

image

image

Comment (minisbett) [reply app/api/domains/cho.py:1236]: fixed

File: app/api/domains/cho.py [hunk 2]
```diff
@@ -1214,9 +1221,16 @@ async def handle(self, player: Player) -> None:
                             # use player mode if not specified
                             mode_vn = player.status.mode.as_vanilla
 
+                        # parse the mods from regex
+                        if r_match["mods"] is not None:
```
Comment (NiceAesth) [inline app/api/domains/cho.py:1225]: i generally don't like usage of if/else blocks like this

nesting is ugly and i try to keep it to a minimum, so i would do
```py
mods = None
if condition:
    mods = val
```

File: app/api/domains/cho.py [hunk 3]
```diff
@@ -397,8 +397,15 @@ async def handle(self, player: Player) -> None:
                         # use player mode if not specified
                         mode_vn = player.status.mode.as_vanilla
 
+                    # parse the mods from regex
+                    if r_match["mods"] is not None:
```
Comment (NiceAesth) [inline app/api/domains/cho.py:401]: same as other one

**Review by NiceAesth — CHANGES_REQUESTED** [review]

File: app/api/domains/cho.py (hunk 2 above)
Comment (minisbett) [reply app/api/domains/cho.py:1225]: well I did it the same way it's on other parts in cho.py, just copied existing if statements and put them in place

File: app/api/domains/cho.py (hunk 3 above)
Comment (minisbett) [reply app/api/domains/cho.py:401]: well I did it the same way it's on other parts in cho.py, just copied existing if statements and put them in place

**Review by cmyui — APPROVED** [review]

**Comment by cmyui** [comment]
+1 👍🏻

**Pull request merged**